### ISSUE TITLE
Use the mainline linux kernel with realtime enabled since it's been merged into mainline

### ISSUE THREAD

**Issue by ivandimitrov8080** [musnix/musnix#206]
PREEMPT_RT has been merged into mainline linux. Use that here please.

**Comment by danwdart** [comment]
It's available "natively" in 6.12.

**Comment by JuxGD** [comment]
Yeah I'm still not sure how to use it. Like if I get 6.12 do I need to do anything in the musnix configuration?

**Comment by hustlerone** [comment]
Set the package to be 6.12

**Comment by henrytill** [comment]
I suspect that even with a 6.12 package, you would still need the extra kernel config - I'd be surprised if `CONFIG_PREEMPT_RT` was enabled by default (I have not confirmed this).  If it is not enabled by default, you will need to modify musnix or provide your own kernel config.

I will have a crack at it soon (contributions welcome!), but I also am interested in eventually using this opportunity to remove our kernel features once and for all, as alluded to in https://github.com/musnix/musnix/issues/118#issuecomment-2351873725.

**Comment by sorathpanzer** [comment]
Sorry for my ignorance, just a humble musician pro-opensource here. What are the benefits of the PREEMPT_RT patch in the kernel vs the regular RT kernel in NiXOS.. Thanks, and keep the good working.

**Comment by ivandimitrov8080** [comment]
> Sorry for my ignorance, just a humble musician pro-opensource here. What are the benefits of the PREEMPT_RT patch in the kernel vs the regular RT kernel in NiXOS.. Thanks, and keep the good working.

Here's what chat gippty thinks about it 

![Image](https://github.com/user-attachments/assets/089cd30a-dae6-400d-a513-c1c44ce47d6b)

Each time I tried to get the RT patch on nixos through this config I had to compile the kernel. Using mainline, I don't have to compile (hopefully) and that way it'll speed up my updates.

**Comment by hustlerone** [comment]
> Sorry for my ignorance, just a humble musician pro-opensource here. What are the benefits of the PREEMPT_RT patch in the kernel vs the regular RT kernel in NiXOS.. Thanks, and keep the good working.

Not using RT kernel for every single damned process.


Also, K900 told me that pipewire runs on real time if you enable rtkit (and on the right kernel)

**Comment by hustlerone** [comment]
iirc RT preemption is already enabled on NixOS

**Comment by AlexLJordan** [comment]
I just switched to `pkgs.linuxPackages_6_12` in hopes of it fixing a graphics driver issue (https://github.com/ROCm/ROCm/issues/2196#issuecomment-2801656114).

There seems to be _some_ improvement, so i also ran rt-test/cyclictest again, but alas the NixOS 6.12 kernel is not PREEMPT_RT, but instead `Linux 6.12.16 #1-NixOS SMP PREEMPT_DYNAMIC Fri Feb 21 13:01:47 UTC 2025`

Now I turned to musnix to get both 6.12 _and_ RT, but as this issue stands, that's not possible yet either?

**Comment by bjsowa** [comment]
To get PREEMPT_RT to work, you need to build the kernel with this option enabled. Here's what worked for me:
```nix
  boot.kernelPackages = pkgs.linuxPackagesFor (pkgs.linux_6_13.override {
    structuredExtraConfig = with lib.kernel; {
      EXPERT = yes;
      PREEMPT_RT = yes;
      RT_GROUP_SCHED = no;
    };
    ignoreConfigErrors = true;
  });
```
(I would appreciate if someone could make it work without  `ignoreConfigErrors` option)

**Comment by Cal-Nine** [comment]
It would be good if musnix.kernel.realtime enabled realtime in other kernels too, e.g. zen, lqx

**Comment by Lord-Valen** [comment]
> It would be good if musnix.kernel.realtime enabled realtime in other kernels too, e.g. zen, lqx

I'm pretty sure those are realtime by default.

**Comment by Cal-Nine** [comment]
> I'm pretty sure those are realtime by default.

You're right! My mistake.

**Comment by hatch01** [comment]
Don't forget to set the kernel to latest or something else !
